File: Binary Search/350. Intersection of Two Arrays II.cpp
/*
Method: This is a solution using binary search method.
        Different from 349. Intersection of Two Arrays, we need to record the appearances of the value as many times as it shows.
        To avoid to save duplicates, we have to remove the value that is saved to the result.
*/
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res;
        sort(nums2.begin(), nums2.end());
        for(auto& val:nums1){
            int pos = binarySearch(nums2, val);
            if(pos!=-1){
                res.push_back(val);
                // vec.erase() takes a position iterator.
                nums2.erase(nums2.begin()+pos);
            }
        }
        return res;
    }
    int binarySearch(vector<int>& nums, int target){
        int left = 0, right = nums.size()-1;
        while(left<=right){
            int mid = left+(right-left)/2;
            if(nums[mid]==target){
                return mid;
            }
            if(nums[mid]>target){
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        return -1;
    }
};

File: Binary Search/350. Intersection of Two Arrays II.py
class Solution(object):
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        """
        Method 1:
        """
        set1 = set(nums1)
        set2 = set(nums2)
        
        intersec = set1.intersection(set2)
        ans = []
        
        for num in intersec:
            min_no = min(nums1.count(num), nums2.count(num))
            for i in range(min_no):
                ans.append(num)
        return ans
        
        
        
        """
        Method 2:
        """
        n1 = collections.Counter(nums1)
        n2 = collections.Counter(nums2)
        n2_sorted = n2.keys()
        n2_sorted.sort()
        
        ans = []
        for n in n1.keys():
            if self.binarySearch(n2_sorted, n):
                times = min(n1[n], n2[n])
                ans.extend([n]*times)
        return ans
    
    def binarySearch(self, arr, target):
        left, right = 0, len(arr)-1
        while left <= right:
            mid = (left+right)//2
            if arr[mid]==target:
                return True
            elif arr[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return False
      
        
